src/blog/pointer-compression.md

```diff
@@ -0,0 +1,544 @@
+
+```cpp
+int32_t compressed_tagged;
+
+// Same code for both pointer and Smi cases
+int64_t sign_extended_tagged = int64_t(compressed_tagged);
+int64_t selector_mask = -(sign_extended_tagged & 1);
+// Mask will be 0 in case of Smi or all 1s in case of pointer
+int64_t uncompressed_tagged =
+    sign_extended_tagged + (base & selector_mask);
+```
+
+Then, we decided to start with the branchless implementation.
+
+## Performance evolution
+
+### Initial performance
+
+We measured performance on [Octane](https://v8.dev/blog/retiring-octane#the-genesis-of-octane) -- a peak-performance benchmark we have used in the past. Although we are no longer focusing on improving peak performance in our day-to-day work, we also don’t want to regress peak performance, particularly for something as performance-sensitive as _all pointers_. Octane continues to be a good benchmark for this task.
+
+This graph shows Octane's score on x64 architecture while we were optimizing and polishing the Pointer Compression implementation. In the graph, higher is better. The red line is the existing full-sized-pointer x64 build, while the green line is the pointer compressed version.
+<figure>
+  <img src="/_img/pointer-compression/perf-octane-1.svg" width="913" height="218" alt="First round of Octane's improvements" loading="lazy">
+  <figcaption>First round of Octane's improvements</figcaption>
+</figure>
+
+With the first working implementation, we had a ~35% regression gap.
+
+#### Bump (1), +7%
+
+First we validated our “branchless is faster” hypothesis, by comparing the branchless decompression with the branchful one. It turned out that our hypothesis was wrong, and the branchful version was 7% faster on x64. That was quite a significant difference!
+
+Let’s take a look at the x64 assembly.
+
+<!-- markdownlint-disable no-inline-html -->
+:::table-wrapper
+<!-- markdownlint-disable no-space-in-code -->
+| Decompression | Branchless              | Branchful                    |
+|---------------|-------------------------|------------------------------|
+| Code          | ```asm                  | ```asm                       \
+|               | movsxlq r11,[...]       | movsxlq r11,[...]            \
+|               | movl r10,r11            | testb r11,0x1                \
+|               | andl r10,0x1            | jz done                      \
+|               | negq r10                | addq r11,r13                 \
+|               | andq r10,r13            | done:                        \
+|               | addq r11,r10            |  ```                         \
+|               | ```                     |  <br>                        |
+| Summary       | 20 bytes                | 13 bytes                     \
+|               | 6 instructions executed | 3 or 4 instructions executed \
+|               | no branches             | 1 branch                     \
+|               | 1 additional register   | <br>                         |
```

Comment:
Wait, so you want all the table cells to be aligned to the top, not just cells with code?

If so, we could add a table class (similarly to `table-wrapper`) that would do just that without any `<br>`s, but it makes table style inconsistent across the blog, as all the other ones are aligned by the middle.

Is there any particular reason you want that alignment?

@mathiasbynens do you have thoughts on this as well maybe?

Reply:
No opinion, other than the usual "let's keep it as simple as possible". I trust you'll make the right decision here 👍

Reply:
Yes, aligning to the top is what we want. Both sides start with loading and then they differ. It is easier to see the difference(e.g in code size) if they are both aligned at the top

Reply:
> No opinion, other than the usual "let's keep it as simple as possible". I trust you'll make the right decision here 👍

I just meant an opinion on concistency. But if you don't have one, that's fine.

Reply:
@santiaboy I see. If the goal is to show clear matching between results, I wonder if it would make more sense to use proper cells (and rowspan for summary here) then?

Here's what it would look like:
![image](https://user-images.githubusercontent.com/557590/77646970-814a6700-6f5d-11ea-90d0-05537a3b42af.png)

You can check table in `v8-release-80.md` for markdown code example.

Reply:
Strong +1 to this suggestion. This seems perfect!

Reply:
Btw, this is the first time we're adding `<pre>` within tables on v8.dev, so let's make sure this works well on narrow viewports etc. before landing.

Reply:
Yes, that looks perfect.

Reply:
> within tables on v8.dev, so let's make sure this works well on narrow viewports etc. before landing.

Yeah I double-checked that already, table-wrapper seems to work fine here too and just introduces horizontal scroll when necessary.